Assets/Plugins/iOS/Shopify/WebCheckoutViewController.swift

```diff
@@ -0,0 +1,236 @@
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import UIKit
+import WebKit
+
+// MARK: View Controller Logic
+class WebCheckoutViewController: UIViewController {
+    weak var delegate: WebCheckoutDelegate?
+    
+    fileprivate(set) var url: URL
+    
+    fileprivate var checkoutView: WebCheckoutView {
+        return view as! WebCheckoutView
+    }
+    
+    fileprivate var webView: WKWebView {
+        return checkoutView.webView
+    }
+    
+    init(url: URL, delegate: WebCheckoutDelegate) {
+        self.url = url
+        self.delegate = delegate
+        
+        super.init(nibName: nil, bundle: nil)
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override func loadView() {
+        self.view = WebCheckoutView(frame: UIScreen.main.bounds)
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        webView.navigationDelegate = self
+        checkoutView.cancelButton.addTarget(self, action: #selector(didPressCancel), for: .touchUpInside)
+        
+        loadCheckoutURL()
+    }
+}
+
+// MARK: Actions
+
+extension WebCheckoutViewController {
+    func loadCheckoutURL() {
+        let request = URLRequest(url: url)
+        webView.load(request)
+        checkoutView.startInitialLoad()
+    }
+    
+    func didPressCancel() {
+        self.delegate?.didCancelCheckout()
+    }
+}
+
+// MARK: WKNavigationDelegate
+extension WebCheckoutViewController: WKNavigationDelegate {
+    
+    public func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {}
+    
+    public func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {}
+    
+    public func webView(_ webView: WKWebView, didCommit navigation: WKNavigation!) {}
+    
+    public func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
+        if checkoutView.initialLoading {
+            // Add a bit of time for the page to render
+            let after: DispatchTime = DispatchTime.now() + .seconds(1)
+            DispatchQueue.main.asyncAfter(deadline: after) {
+                self.checkoutView.endInitialLoad()
+            }
+            
+            return
+        }
+        
+        if let url = webView.url, isThankYou(url: url) {
+            // Change the color and title of the button
+            checkoutView.setButton(state: .finished)
+            delegate?.didFinishCheckout()
```

Comment:
Instead of a single `didCloseWebView` method I split it up into two calls: `didFinishCheckout` and `didCancelCheckout`. My thoughts are that maybe the game developer would want to know when a user completes a purchase vs cancelling out of the checkout process.

The more I think about it, it might make sense to have a single callback for when the web view dismissed (`didDismissWebCheckout`) and have another one specifically for when the user completes a purchase `didCompletePurchase` which is fired when we get on the thank_you page.